### PR TITLE
Allow to specify a custom EventExecutorChooserFactory. Related to [#1…

### PR DESCRIPTION
**Pull request by normanmaurer**
…230]

Motivation:

Sometimes it may be benefitially for an user to specify a custom algorithm when choose the next EventExecutor/EventLoop.

Modifications:

Allow to specify a custom EventExecutorChooseFactory that allows to customize algorithm.

Result:

More flexible api.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin upstream

File: common/src/main/java/io/netty/util/concurrent/EventExecutorChooserFactory.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Factory that creates new {@link EventExecutorChooser}s.
+ */
+public interface EventExecutorChooserFactory {
+
+    /**
+     * Returns a new {@link EventExecutorChooser}.
+     */
+    EventExecutorChooser newChooser(EventExecutor[] executors);
```
Comment (netkins) [inline common/src/main/java/io/netty/util/concurrent/EventExecutorChooserFactory.java]: MINOR Consider using varargs for methods or constructors which take an array the last parameter. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    private static boolean isPowerOfTwo(int val) {
+        return (val & -val) == val;
+    }
+
+    private static final class PowerOfTowEventExecutorChooser implements EventExecutorChooser {
+        private final AtomicInteger idx = new AtomicInteger();
+        private final EventExecutor[] executors;
+
+        PowerOfTowEventExecutorChooser(EventExecutor[] executors) {
```
Comment (netkins) [inline common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: MINOR Consider using varargs for methods or constructors which take an array the last parameter. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        public EventExecutor next() {
+            return executors[idx.getAndIncrement() & executors.length - 1];
+        }
+    }
+
+    private static final class GenericEventExecutorChooser implements EventExecutorChooser {
+        private final AtomicInteger idx = new AtomicInteger();
+        private final EventExecutor[] executors;
+
+        GenericEventExecutorChooser(EventExecutor[] executors) {
```
Comment (netkins) [inline common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: MINOR Consider using varargs for methods or constructors which take an array the last parameter. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ */
+public final class DefaultEventExecutorChooserFactory implements EventExecutorChooserFactory {
+
+    public static final DefaultEventExecutorChooserFactory INSTANCE = new DefaultEventExecutorChooserFactory();
+
+    private DefaultEventExecutorChooserFactory() { }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public EventExecutorChooser newChooser(EventExecutor[] executors) {
```
Comment (netkins) [inline common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: MINOR Consider using varargs for methods or constructors which take an array the last parameter. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        } else {
+            return new GenericEventExecutorChooser(executors);
+        }
+    }
+
+    private static boolean isPowerOfTwo(int val) {
+        return (val & -val) == val;
+    }
+
+    private static final class PowerOfTowEventExecutorChooser implements EventExecutorChooser {
```
Comment (johnou) [inline common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: Power of tow? :D

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        private final AtomicInteger idx = new AtomicInteger();
+        private final EventExecutor[] executors;
+
+        PowerOfTowEventExecutorChooser(EventExecutor[] executors) {
+            this.executors = executors;
+        }
+
+        @Override
+        public EventExecutor next() {
+            return executors[idx.getAndIncrement() & executors.length - 1];
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: save the mask in a `private final`?

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    public static final DefaultEventExecutorChooserFactory INSTANCE = new DefaultEventExecutorChooserFactory();
+
+    private DefaultEventExecutorChooserFactory() { }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public EventExecutorChooser newChooser(EventExecutor[] executors) {
+        if (isPowerOfTwo(executors.length)) {
+            return new PowerOfTowEventExecutorChooser(executors);
+        } else {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: nit: remove `else`

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            return new GenericEventExecutorChooser(executors);
+        }
+    }
+
+    private static boolean isPowerOfTwo(int val) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: consider moving this to MathUtil

**Comment by Scottmitch** [comment]
@normanmaurer - did you close this intentionally?

**Pull request closed without merging**